articles/contributing/api-design-principles.md

```diff
@@ -0,0 +1,426 @@
+        implementation details.
+
+- **DO** choose operation and function names that reflect the
+    high-level structure of algorithms and applications.
+
+- **DO** start each API design with sample use cases to ensure that
+    APIs are intuitive to use.
+  - **DO** ensure that each component of a public API has a
+      corresponding use case, rather than trying to design for all possible uses from the start.
+      Put differently, don't introduce public APIs in
+      case they are useful, but make sure that each part of an API has
+      a *concrete* example in which it will be useful. For example:
+    - @"microsoft.quantum.canon.applytoeachca" can be used as `ApplyToEachCA(H, _)` to prepare
+        registers in a uniform superposition state, a common task in
+        many quantum algorithms. The same operation can also be used
+        for many other tasks in preparation, numerics, and
+        oracle-based algorithms.
+  - **DO** brainstorm and workshop new API designs to double-check
+      that they are intuitive and meet proposed use cases. For
+      example:
+    - Inspect current Q\# code to see how new API designs could
+        simplify and clarify existing implementations.
+    - Review proposed API designs with representatives of primary
+        audiences.
+
+- **DO** design APIs to support and encourage readable code.
+  - **DO** ensure that code is readable by domain experts and
+      non-experts alike.
+  - **DO** place the focus on the effects of each operation and
+      function within the high-level algorithm, using documentation to
+      delve into implementation details as appropriate.
+
+- **DO** design APIs to be stable and to provide forward
+    compatibility.
+
+- **DO** deprecate old APIs gracefully when breaking changes are
+    required.
+
+  - **DO** provide "shim" operations and functions that allow
+      existing user code to operate correctly during deprecation. For
+      example:
+    - When renaming an operation called `EstimateExpectation` to
+        `EstimateAverage`, introduce a new operation called
+        `EstimateExpectation` that calls the original operation at
+        its new name, so that existing code can continue to work
+        correctly.
+  - **DO** use the @"microsoft.quantum.core.deprecated" attribute to communicate
+      deprecations to the user.
+  - When renaming an operation or function, **DO** provide the new
+      name as a string input to `@Deprecated`.
+  - **DON'T** remove existing functions or operations without a
+      deprecation period of at least six months for preview releases,
+      or at least two years for supported releases.
+
+- **DO** follow the common [Q\# style guide](xref:microsoft.quantum.contributing.style)
+    whenever applicable.
+
+## Functions and Operations
+
+- **DO** ensure that every function and operation has a single
+    well-defined purpose within the API.
+  - **DON'T** expose functions and operations that perform multiple
```

Comment:
I'm not sure having indented do's and don'ts is a good idea. We should find another way to group them.

Reply:
One thought might be to have each **do** and **don't** under a corresponding key principles:

![image](https://user-images.githubusercontent.com/31516/75284999-a2872e80-57ca-11ea-9509-14231fecbbec.png)

Reply:
I really like that format, could also then have a slightly different style under the key principle for examples in the QDK that follow that principle?

Reply:
That's what I was thinking, @crazy4pi314. It will take a bit to rework to that format, but I'll get started on that and work to incorporate additional examples as per suggestions from you and @anpaz-msft.

Reply:
I've refactored to the suggested format, adding key principle headings, ✅  and ⛔️, separate example headings and so forth to help emphasize. Please feel free to re-open if you feel there's more work to do here. Thanks!